core.go

```diff
@@ -53,9 +54,49 @@ func (c Core) ListObjectsV2(bucketName, objectPrefix, continuationToken string,
 	return c.listObjectsV2Query(bucketName, objectPrefix, continuationToken, fetchOwner, delimiter, maxkeys)
 }
 
+// CopyObject - copies an object from source object to destination object on server side.
+func (c Core) CopyObject(sourceBucket, sourceObject, destBucket, destObject string, metadata map[string]string) (ObjectInfo, error) {
+	return c.copyObjectDo(context.Background(), sourceBucket, sourceObject, destBucket, destObject, metadata)
+}
+
 // PutObject - Upload object. Uploads using single PUT call.
 func (c Core) PutObject(bucket, object string, data io.Reader, size int64, md5Sum, sha256Sum []byte, metadata map[string]string) (ObjectInfo, error) {
-	return c.putObjectDo(context.Background(), bucket, object, data, md5Sum, sha256Sum, size, PutObjectOptions{UserMetadata: metadata})
+	delKey := func(m map[string]string, key string) {
+		delete(m, strings.ToLower(key))
+		delete(m, strings.ToUpper(key))
+	}
+
+	getKeyValue := func(m map[string]string, key string) string {
+		v, ok := m[strings.ToLower(key)]
+		if ok {
+			return v
+		}
+		v, ok = m[strings.ToUpper(key)]
+		if ok {
+			return v
+		}
+		return ""
+	}
+
+	opts := PutObjectOptions{}
+	if v := getKeyValue(metadata, "cache-control"); v != "" {
+		opts.CacheControl = v
+		delKey(metadata, "cache-control")
+	}
+	if v := getKeyValue(metadata, "content-type"); v != "" {
+		opts.ContentType = v
+		delKey(metadata, "content-type")
+	}
+	if v := getKeyValue(metadata, "content-disposition"); v != "" {
+		opts.ContentDisposition = v
+		delKey(metadata, "content-disposition")
+	}
+	if v := getKeyValue(metadata, "content-encoding"); v != "" {
+		opts.ContentEncoding = v
+		delKey(metadata, "content-encoding")
+	}
```

Comment:
```
opts := PutObjectOptions{}
m := make(map[string]string)
for k, v := range metadata {
	if strings.ToLower(k) == "content-encoding" {
		opts.ContentEncoding = v
	} else if strings.ToLower(k) == "content-disposition" {
		opts.ContentDisposition = v
	} else if strings.ToLower(k) == "content-type" {
		opts.ContentType = v
	} else if strings.ToLower(k) == "cache-control" {
		opts.CacheControl = v
	} else {
		m[k] = metadata[k]
	}
}
opts.UserMetadata = m
```
this would be more concise

Reply:
This makes sense..